src/metabase/models/serialization/base.clj

```diff
@@ -0,0 +1,171 @@
+
+    Only called by `serialize-all`, either the default implementation or (optionally) by a custom one.
+    This exists to be an easy override point, when the default `serialize-all` is good but you need to filter
+    the returned set, eg. by dropping archived entities.")
+
+  (serialize-one
+    [this]
+    "Serializes a single entity into a YAML-friendly map, with its filename.
+    Returns a [file-name {entity map...}] pair.
+
+    Default implementation:
+    - Uses `\"$ENTITY_ID.yaml\"` or `\"$IDENTITY_HASH.yaml\"` as the filename.
+    - Returns the entity as a vanilla map with `:id` and any `:foo_at` fields removed.
+    - Adds the field `:type \"Collection\"` (etc.) to specify the model.
+
+    That suffices for a few simple entities, but most entities will need to override this.
+    They should follow the pattern of dropping timestamps and numeric database IDs, and including the `:type`.")
+
+  ;;; Deserialization side.
+  (deserialize-file
+    [this new-map local]
+    "Given the model, the file contents as an EDN map, and any corresponding entity already existing in the database,
+    this performs whatever steps are necessary to update or insert the incoming entity.
+    Called for each **file**, which usually means for each entity. For single-file output like for settings, this is
+    called once for the whole file.
+    Defaults to [[deserialize-file-plus]] with [[identity]] as the inner transformation.")
+
+  (deserialize-upsert
+    [old-entity new-map]
+    "Given the original entity and the new deserialized map (not an IModel yet), upsert appdb to merge in the updated
+    entity.
+    Defaults to a naive update by primary key, using just the new map's values.")
+
+  (deserialize-insert
+    [this new-map]
+    "Given the model (eg. Collection) and the new deserialized map (not an IModel), insert this new record into
+    the appdb.
+    Defaults to a straight db/insert! of this new map."))
+
+(defn- default-serialize-all [model user-or-nil]
+  (eduction (map serialize-one) (serialize-query model user-or-nil)))
+
+(defn- default-query [model _]
+  (db/select-reducible model))
+
+(defn serialize-one-plus
+  "Helper for applying the usual serialization steps to a single entity, followed by a user-supplied function with any
+  specific logic the entity needs."
+  [transform named]
+  (fn [entity]
+    [(format "%s%s%s%s.yaml"
+             (name entity)
+             File/separatorChar
+             (or (:entity_id entity)
+                 (serdes.hash/identity-hash entity))
+             (when named
+               (str "+" (named entity))))
+     (-> (into {} entity)
+         (dissoc (models/primary-key entity))
+         (assoc :serdes_type (name entity))
+         transform)]))
+
+(defn- default-upsert [old-entity new-map]
+  (db/update! (symbol (name old-entity)) (get old-entity (models/primary-key old-entity)) new-map)
+  #_(prn "upsert" old-entity new-map))
```

Comment:
I would change these to `log/trace` or `log/debug` and uncomment them so you can see this stuff while working on it.

```clj
(metabase.test/set-ns-log-level! 'metabase.models.serialization.base :trace)
```

is a tool I wrote a while back that will let you flip the log level of a specific namespace from the REPL.

Reply:
Oh, cool tool.

Those are actually leftover debugging code from when I didn't want to change my dev appdb. I'll turn them into permanent `log/debug`s.